cdap-app-fabric/src/test/java/co/cask/cdap/scheduler/CoreSchedulerServiceTest.java

```diff
@@ -320,6 +339,63 @@ public boolean apply(Job job) {
     testScheduleUpdate("delete");
   }
 
+  @Test
+  @Category(XSlowTests.class)
+  public void testProgramEvents() throws Exception {
+    // Deploy the app
+    deploy(AppWithMultipleSchedules.class);
+    CConfiguration cConf = CConfiguration.create();
+
+    TopicId programEventTopic = NamespaceId.SYSTEM.topic(cConf.get(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC));
+    ProgramStateWriter programStateWriter = new MessagingProgramStateWriter(cConf, messagingService);
+
+    long lastProcessed = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
+
+    // These notifications should not trigger the program
+    ProgramRunId anotherWorkflowRun = ANOTHER_WORKFLOW.run(RunIds.generate());
+    programStateWriter.start(anotherWorkflowRun, new SimpleProgramOptions(anotherWorkflowRun.getParent()), null);
+    programStateWriter.running(anotherWorkflowRun, null);
+    programStateWriter.error(anotherWorkflowRun, null);
+    waitUntilProcessed(programEventTopic, lastProcessed);
+    Assert.assertEquals(0, getRuns(TRIGGERED_WORKFLOW));
+
+    lastProcessed = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
+    ProgramRunId someWorkflowRun = SOME_WORKFLOW.run(RunIds.generate());
+    programStateWriter.start(someWorkflowRun, new SimpleProgramOptions(someWorkflowRun.getParent()), null);
+    programStateWriter.running(someWorkflowRun, null);
+    programStateWriter.killed(someWorkflowRun);
```

Comment:
get the time for `lastProcessed` above this line, so that the time is closest to the last message's publishing